### ISSUE TITLE
CMOR 3.2.7 didn't apply valid_min and valid_max attributes to obs4MIPs output variable

### ISSUE THREAD

**Issue by JimBiardCics** [PCMDI/cmor#237]
I generated an obs4MIPs file for ndvi (in Lmon table). The table contains valid_min and valid_max attributes for the variable. CMOR warned that some values were below the valid_min (which is by design - some values are invalid but serve a purpose for helping advanced users understand the limitations of the NDVI algorithm used). After warning that some values were out of the valid range, the output file from CMOR did not contain the valid_min or valid_max attributes on the ndvi variable.

**Comment by durack1** [comment]
This might be helpful background [PCMDI/obs4MIPs-cmor-tables/Tables/obs4MIPs_Lmon.json#ndvi](https://github.com/PCMDI/obs4MIPs-cmor-tables/blob/master/Tables/obs4MIPs_Lmon.json#L630-L646)

**Comment by taylor13** [comment]
I don't think any modifications of CMOR are needed.  This is exactly what we want CMOR to do.

**Comment by JimBiardCics** [comment]
@taylor13 The valid_min and valid_max attributes have a purpose in CF. They are useful in a netCDF file in general, and would be useful in our ndvi obs4MIPs files in particular. As it stands, CMOR generates a warning based on comparisons done with the attribute values, but it neither removes the 'out of range' values, nor passes the attributes on to the output. This leaves the output obs4MIPs file in a bad state as far as I'm concerned. Why is this what you want CMOR to do?

**Comment by taylor13** [comment]
understood.   In CMIP, valid_min and valid_max are meant to trap suspected gross errors in the input data.  For example, suppose a user says he is passing temperatures in degrees Celsius to CMOR, but CMOR detects values of order 300 degrees.   Since 300 presumably is much larger than valid_max for an atmospheric temperature expressed in Celsius, CMOR would raise a warning.  Now it is possible that a very unrealistic model would produce temperatures that high for the atmosphere (or perhaps extend into the exosphere), in which case the data provider could ignore the warning (or improve his model). 

If you attach special meaning to data values that are not within a valid range (in the CF convention sense), you can always ask CMOR to write a "valid_min" and "valid_max" attribute to the file attached to your variable.  You do this with cmor_set_variable_attribute (see https://cmor.llnl.gov/mydoc_cmor3_api/#cmor_set_variable_attribute).

**Comment by JimBiardCics** [comment]
@taylor13 I have no argument with using the valid_min and valid_max values specified in the tables in the way that you are doing. I'm suggesting that you should **also** put the attributes on the output variable. I'm happy to get the warning. It just seems counterintuitive to specify the attributes in the tables and not have them show up in the output.

**Comment by taylor13** [comment]
The trouble is that in CMIP the value of a variable can be outside the "valid_min" and "valid_max" range and still be valid.  We should have called these limits some different from their CF names (e.g. approx_upper_limit and approx_lower_value).  I would see little value in storing approx_upper_limit and approx_lower_value in the netCDF files, and it would incorrect to store their values in valid_min and valid_max.

**Comment by JimBiardCics** [comment]
@taylor13 Ah. That is, indeed, quite unfortunate. In addition, CMOR isn't propagating valid_min or valid_max attributes from within the input variable, so I'm stuck with special-casing if I want to get these values into my output file. @dnadeau4 Why aren't these attributes inherited from the input file? They are just about as important for interpreting values in a variable as _FillValue and missing_value.

**Comment by taylor13** [comment]
As note above in https://github.com/PCMDI/cmor/issues/237#issuecomment-332666358:

If you attach special meaning to data values that are not within a valid range (in the CF convention sense), you can always ask CMOR to write a "valid_min" and "valid_max" attribute to the file attached to your variable. You do this with cmor_set_variable_attribute (see https://cmor.llnl.gov/mydoc_cmor3_api/#cmor_set_variable_attribute).

Let me try again: The reason they aren't inherited is because valid_min and valid_max (in the sense of the CF conventions) are not defined in the table; only approx_lower_value and approx_upper_value are in the table (but they have the names valid_min and valid_max).

**Comment by JimBiardCics** [comment]
I get the problem with using the unfortunately named valid_min and valid_max from the Tables. But that is independent of the question of why you aren't propagating actual CF valid_min and valid_max attributes from the input file variable into the output.

**Comment by dnadeau4** [comment]
I looked at the code and `valid_min`, `valid_max`,  `ok_min_means_abs`, `ok_max_means_abs`, `shuffle`, `deflate`, `deflate_level` are not set as variable attributes.   That decision was made before CMIP5.  I understand the reason for some of them.  Others are not CF-1 compliant.

**Comment by JimBiardCics** [comment]
Only valid_min and valid_max are CF attributes.

**Comment by taylor13** [comment]
The valid_min and valid_max values (in the CF sense) would have to be defined by each data provider.  We can't guess what the range of their valid data are.  The point of CMOR is to put into tables things that will be same for all people using CMOR who rely on a given table.

**Comment by JimBiardCics** [comment]
@taylor13 What do you suggest then? As a data provider, I have defined the values. They are in my input files. I'm fine with not having the values for my dataset in the Tables JSON files. We've resolved the question of why the entries in the Tables aren't what I had originally expected.

What I'm saying is that as a CMOR user I am surprised that significant CF attributes on my input variables that provide important information to downstream users - specifically valid_min and valid_max (which are not the same as the ones in the Tables) - are dropped by CMOR rather than transferred to my output variables.

**Comment by durack1** [comment]
@JimBiardCics, while it's not exactly what you are requesting, you can assign new global attributes by setting these in the [`userInput.json` (this links to the demo example)](https://github.com/PCMDI/obs4MIPs-cmor-tables/blob/master/demo/rssSsmiPrw-input.json), so
```
...
"#comment": "The user may choose to define or omit/delete some or all of the following",
...
"references": "",
"valid_min": "2.0",
"valid_max": "3.0",
...
```
Will yield (in an `ncdump -h`):
```
netcdf prw_mon_REMSS-PRW-6-6-0_BE_gn_198701-198812 {
dimensions:
	time = UNLIMITED ; // (24 currently)
	lat = 72 ;
	lon = 144 ;
	bnds = 2 ;
variables:
...
// global attributes:
...
:tracking_id = "hdl:21.14102/5b51e145-ca14-4cb8-92e6-3c290341c052" ;
:valid_max = "3.0" ;
:valid_min = "2.0" ;
:variable_id = "prw" ;
...
```

**Comment by JimBiardCics** [comment]
@durack1 That doesn't help. The attributes must be on the variable to be meaningful.

**Comment by taylor13** [comment]
As noted in https://github.com/PCMDI/cmor/issues/237#issuecomment-332666358 and again in https://github.com/PCMDI/cmor/issues/237#issuecomment-332899250, you can easily set variable attributes with cmor_set_variable_attribute (see https://cmor.llnl.gov/mydoc_cmor3_api/#cmor_set_variable_attribute).
